File: reaper/src/protocol.rs
//! The reaper protocol: one JSON object per line (NDJSON).
//!
//! The host (the process that owns the session) writes [`Request`]s to the
//! reaper's stdin; the reaper writes [`Event`]s to its stdout. The TypeScript
//! side of this contract is `src/reaper/protocol.ts`; change both together.
//!
//! | Host → reaper | Meaning |
//! | ------------- | ------- |
//! | `go`          | Admission: the reaper may spawn from now on. |
//! | `spawn`       | Start one worker in its own job (Windows) or process group (POSIX). |
//! | `stop`        | Stop one worker: graceful signal, then forced kill of its tree after `graceMs`. |
//! | `terminate`   | Irreversible end: reject later spawns, stop every worker, exit. |
//!
//! | Reaper → host | Meaning |
//! | ------------- | ------- |
//! | `leased`      | The reaper holds the lease; it waits for `go`. |
//! | `spawned`     | A worker exists (PID and OS start time). |
//! | `rejected`    | A `spawn` did not start a worker. |
//! | `exited`      | A worker's whole tree is gone (or the bound ran out). |
//! | `terminated`  | Every tree is empty; the reaper releases the lease and exits. |

use std::collections::BTreeMap;

use serde::{Deserialize, Serialize};

/// One worker to start.
#[derive(Clone, Debug, Deserialize, Eq, PartialEq)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
pub struct SpawnRequest {
    /// The host's name for the worker, unique among live workers.
    pub id: String,
    /// Absolute path of the executable.
    pub file: String,
    pub args: Vec<String>,
    pub cwd: String,
    /// The complete environment of the worker.
    pub env: BTreeMap<String, String>,
    /// File that receives stdout and stderr (appended). No file: discarded.
    #[serde(default)]
    pub log: Option<String>,
    /// Windows only: join `args` with spaces, without quoting (cmd.exe
    /// command lines that are already escaped).
    #[serde(default)]
    pub verbatim: bool,
}

/// A message from the host.
#[derive(Clone, Debug, Deserialize, Eq, PartialEq)]
#[serde(deny_unknown_fields, rename_all = "camelCase", tag = "type")]
pub enum Request {
    Go,
    Spawn(SpawnRequest),
    #[serde(rename_all = "camelCase")]
    Stop {
        id: String,
        grace_ms: u64,
    },
    #[serde(rename_all = "camelCase")]
    Terminate {
        grace_ms: u64,
    },
}

/// How one worker ended.
#[derive(Clone, Debug, Default, Eq, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct Report {
    /// The leader's exit code; `null` when a signal ended it (POSIX).
    pub exit_code: Option<i32>,
    /// The signal number that ended the leader (POSIX only).
    pub signal: Option<i32>,
    /// The reaper had to kill the tree: the graceful stop did not empty it
    /// in time, or members outlived the leader.
    pub forced: bool,
    /// Members were still alive when the confirmation bound ran out.
    pub incomplete: bool,
}

/// Why a `spawn` started no worker.
#[derive(Clone, Copy, Debug, Eq, PartialEq, Serialize)]
#[serde(rename_all = "snake_case")]
pub enum RejectReason {
    /// `go` has not arrived.
    NotAdmitted,
    /// `terminate` (or stdin EOF) came first; the reaper never spawns again.
    Terminating,
    /// A live worker has this id.
    DuplicateId,
    /// The OS could not start the process.
    SpawnFailed,
}

/// One entry of `terminated`.
#[derive(Clone, Debug, Eq, PartialEq, Serialize)]
pub struct FinalReport {
    pub id: String,
    pub report: Report,
}

/// A message to the host.
#[derive(Clone, Debug, Eq, PartialEq, Serialize)]
#[serde(rename_all = "camelCase", tag = "type")]
pub enum Event {
    Leased {
        pid: u32,
    },
    #[serde(rename_all = "camelCase")]
    Spawned {
        id: String,
        pid: u32,
        /// OS start time as a decimal string (see `os::process`).
        start_time: String,
    },
    Rejected {
        id: String,
        reason: RejectReason,
        message: String,
    },
    Exited {
        id: String,
        report: Report,
    },
    /// Every worker the reaper ever spawned, in spawn order.
    Terminated {
        reports: Vec<FinalReport>,
    },
}

/// Read one request line.
///
/// # Errors
///
/// When the line is not a valid request.
pub fn parse_request(line: &str) -> serde_json::Result<Request> {
    serde_json::from_str(line)
}

/// Write one event as a line, newline included.
#[must_use]
pub fn encode_event(event: &Event) -> String {
    let mut line = serde_json::to_string(event).expect("events always serialize");
    line.push('\n');
    line
}

#[cfg(test)]
mod tests {
    use super::{Event, RejectReason, Report, Request, SpawnRequest, encode_event, parse_request};
    use std::collections::BTreeMap;

    #[test]
    fn parses_every_request() {
        assert_eq!(parse_request(r#"{"type":"go"}"#).unwrap(), Request::Go);
        assert_eq!(
            parse_request(r#"{"type":"stop","id":"rojo","graceMs":250}"#).unwrap(),
            Request::Stop {
                id: "rojo".into(),
                grace_ms: 250
            }
        );
        assert_eq!(
            parse_request(r#"{"type":"terminate","graceMs":0}"#).unwrap(),
            Request::Terminate { grace_ms: 0 }
        );
    }

    #[test]
    fn parses_a_spawn_with_defaults() {
        let line = r#"{"type":"spawn","id":"rojo","file":"/bin/rojo","args":["serve"],"cwd":"/p","env":{"A":"1"}}"#;
        assert_eq!(
            parse_request(line).unwrap(),
            Request::Spawn(SpawnRequest {
                id: "rojo".into(),
                file: "/bin/rojo".into(),
                args: vec!["serve".into()],
                cwd: "/p".into(),
                env: BTreeMap::from([("A".into(), "1".into())]),
                log: None,
                verbatim: false,
            })
        );
    }

    #[test]
    fn rejects_unknown_types_and_bad_json() {
        assert!(parse_request(r#"{"type":"explode"}"#).is_err());
        assert!(parse_request("not json").is_err());
    }

    #[test]
    fn encodes_events_as_camel_case_lines() {
        assert_eq!(
            encode_event(&Event::Spawned {
                id: "rojo".into(),
                pid: 7,
                start_time: "99".into()
            }),
            "{\"type\":\"spawned\",\"id\":\"rojo\",\"pid\":7,\"startTime\":\"99\"}\n"
        );
        assert_eq!(
            encode_event(&Event::Rejected {
                id: "a".into(),
                reason: RejectReason::NotAdmitted,
                message: "m".into()
            }),
            "{\"type\":\"rejected\",\"id\":\"a\",\"reason\":\"not_admitted\",\"message\":\"m\"}\n"
        );
        assert_eq!(
            encode_event(&Event::Exited {
                id: "a".into(),
                report: Report {
                    exit_code: Some(0),
                    ..Report::default()
                }
            }),
            "{\"type\":\"exited\",\"id\":\"a\",\"report\":{\"exitCode\":0,\"signal\":null,\"forced\":false,\"incomplete\":false}}\n"
        );
    }
}

File: src/reaper/protocol.ts
import type { Type } from "arktype";
import { type } from "arktype";

import type { Environment } from "../seams/seams.ts";

/**
 * The reaper protocol, host side: one JSON object per line. The host writes
 * {@link ReaperRequest}s to the reaper's stdin and reads
 * {@link ReaperEvent}s from its stdout. The Rust side is
 * `reaper/src/protocol.rs`; change both together.
 */

/** One worker for the reaper to start. */
export interface WorkerSpec {
	/** The host's name for the worker, unique among live workers. */
	id: string;
	args: ReadonlyArray<string>;
	cwd: string;
	/** The worker's whole environment; the reaper adds the markers. */
	env: Environment;
	/** Absolute path of the executable. */
	file: string;
	/** File that gets stdout and stderr, appended. None: discarded. */
	log?: string;
	/** Windows only: pass `args` joined, unquoted (escaped cmd.exe lines). */
	verbatimArguments?: boolean;
}

/** A message to the reaper. */
export type ReaperRequest =
	| { graceMs: number; id: string; type: "stop" }
	| { graceMs: number; type: "terminate" }
	| { type: "go" }
	| { type: "spawn"; worker: WorkerSpec };

/** How one worker's tree ended. */
export interface WorkerReport {
	/** The leader's exit code; `null` when a signal ended it. */
	exitCode: null | number;
	/** The reaper killed the tree: the graceful stop ran out of time. */
	forced: boolean;
	/** Members still lived when the reaper stopped waiting for them. */
	incomplete: boolean;
	/** The POSIX signal number that ended the leader. */
	signal: null | number;
	/**
	 * POSIX: the PIDs of the members that still lived. The reaper always
	 * sends the list (empty on Windows, which cannot name them); a report the
	 * host makes for a dead reaper has none.
	 */
	survivors?: Array<number>;
}

/** Why the reaper started no worker for a `spawn`. */
export type RejectReason = "duplicate_id" | "not_admitted" | "spawn_failed" | "terminating";

/** One worker in `terminated`. */
export interface FinalReport {
	id: string;
	report: WorkerReport;
}

/** A message from the reaper. */
export type ReaperEvent =
	| { id: string; message: string; reason: RejectReason; type: "rejected" }
	| { id: string; pid: number; startTime: string; type: "spawned" }
	| { id: string; report: WorkerReport; type: "exited" }
	| { pid: number; type: "leased" }
	| { reports: Array<FinalReport>; type: "terminated" };

const report = type({
	"exitCode": "number.integer | null",
	"forced": "boolean",
	"incomplete": "boolean",
	"signal": "number.integer | null",
	"survivors?": "number.integer[]",
});

const eventSchema: Type<ReaperEvent> = type.or(
	{ pid: "number.integer", type: "'leased'" },
	{ id: "string", pid: "number.integer", startTime: "string", type: "'spawned'" },
	{
		id: "string",
		message: "string",
		reason: "'duplicate_id' | 'not_admitted' | 'spawn_failed' | 'terminating'",
		type: "'rejected'",
	},
	{ id: "string", report, type: "'exited'" },
	{ reports: type({ id: "string", report }).array(), type: "'terminated'" },
);

const eventLine = type("string.json.parse").pipe(eventSchema);

/**
 * Write one request as a line.
 *
 * @param request - What to send.
 * @returns The JSON text and a newline.
 */
export function encodeRequest(request: ReaperRequest): string {
	if (request.type !== "spawn") {
		return `${JSON.stringify(request)}\n`;
	}

	const { worker } = request;
	const line = {
		id: worker.id,
		args: worker.args,
		cwd: worker.cwd,
		env: definedVariables(worker.env),
		file: worker.file,
		log: worker.log,
		type: "spawn",
		verbatim: worker.verbatimArguments === true,
	};
	return `${JSON.stringify(line)}\n`;
}

/**
 * Read one line from the reaper.
 *
 * @param line - One line of its stdout, without the newline.
 * @returns The event, or `undefined` when the line is not one.
 */
export function parseEvent(line: string): ReaperEvent | undefined {
	const event = eventLine(line);
	return event instanceof type.errors ? undefined : event;
}

function definedVariables(environment: Environment): Record<string, string> {
	const defined: Record<string, string> = {};
	for (const [name, value] of Object.entries(environment)) {
		if (value !== undefined) {
			defined[name] = value;
		}
	}

	return defined;
}
